File: CryptoSavingsTracker/Views/Components/ImpactPreviewCard.swift
//
//  ImpactPreviewCard.swift
//  CryptoSavingsTracker
//
//  Created by Claude on 07/08/2025.
//

import SwiftUI

struct ImpactPreviewCard: View {
    let impact: GoalImpact
    let currency: String
    
    @State private var animatedOldProgress: Double = 0
    @State private var animatedNewProgress: Double = 0
    
    init(impact: GoalImpact, currency: String = "USD") {
        self.impact = impact
        self.currency = currency
    }
    
    var body: some View {
        VStack(spacing: 16) {
            // Header
            HStack {
                Image(systemName: impact.isPositiveChange ? "arrow.up.right.circle.fill" : "arrow.down.right.circle.fill")
                    .foregroundColor(impact.isPositiveChange ? AccessibleColors.success : AccessibleColors.warning)
                    .font(.title2)
                
                VStack(alignment: .leading, spacing: 2) {
                    Text("Impact Preview")
                        .font(.headline)
                        .fontWeight(.medium)
                    
                    Text(impact.isPositiveChange ? "Positive change" : "Requires attention")
                        .font(.caption)
                        .foregroundColor(impact.isPositiveChange ? AccessibleColors.success : AccessibleColors.warning)
                }
                
                Spacer()
            }
            
            // Progress Comparison
            VStack(spacing: 12) {
                HStack {
                    Text("Progress")
                        .font(.subheadline)
                        .fontWeight(.medium)
                    Spacer()
                }
                
                HStack(spacing: 20) {
                    // Before
                    VStack(spacing: 8) {
                        Text("Before")
                            .font(.caption)
                            .foregroundColor(.accessibleSecondary)
                        
                        CircularProgressView(
                            progress: animatedOldProgress,
                            size: 60,
                            lineWidth: 6,
                            color: AccessibleColors.chartColor(at: 1)
                        )
                        
                        Text("\(Int(impact.oldProgress * 100))%")
                            .font(.caption)
                            .fontWeight(.medium)
                            .contentTransition(.numericText())
                    }
                    
                    // Arrow
                    Image(systemName: "arrow.right")
                        .foregroundColor(.accessibleSecondary)
                        .font(.title2)
                    
                    // After
                    VStack(spacing: 8) {
                        Text("After")
                            .font(.caption)
                            .foregroundColor(.accessibleSecondary)
                        
                        CircularProgressView(
                            progress: animatedNewProgress,
                            size: 60,
                            lineWidth: 6,
                            color: impact.isPositiveChange ? AccessibleColors.success : AccessibleColors.warning
                        )
                        
                        Text("\(Int(impact.newProgress * 100))%")
                            .font(.caption)
                            .fontWeight(.medium)
                            .contentTransition(.numericText())
                    }
                }
            }
            
            // Key Changes
            VStack(spacing: 8) {
                if abs(impact.targetAmountChange) > 0.01 {
                    ChangeRow(
                        title: "Target Amount",
                        oldValue: String(format: "%.2f %@", impact.oldTargetAmount, currency),
                        newValue: String(format: "%.2f %@", impact.newTargetAmount, currency),
                        isPositive: impact.targetAmountChange > 0
                    )
                }
                
                if abs(impact.dailyTargetChange) > 0.01 {
                    ChangeRow(
                        title: "Daily Target",
                        oldValue: String(format: "%.2f %@", impact.oldDailyTarget, currency),
                        newValue: String(format: "%.2f %@", impact.newDailyTarget, currency),
                        isPositive: impact.dailyTargetChange < 0 // Less daily target is better
                    )
                }
                
                if impact.daysRemainingChange != 0 {
                    ChangeRow(
                        title: "Days Remaining",
                        oldValue: "\(impact.oldDaysRemaining) days",
                        newValue: "\(impact.newDaysRemaining) days",
                        isPositive: impact.daysRemainingChange > 0 // More days is better
                    )
                }
            }
            
            // Warning if significant negative change
            if !impact.isPositiveChange && impact.significantChange {
                HStack {
                    Image(systemName: "exclamationmark.triangle.fill")
                        .foregroundColor(AccessibleColors.warning)
                    
                    VStack(alignment: .leading, spacing: 2) {
                        Text("Significant Change")
                            .font(.caption)
                            .fontWeight(.medium)
                        
                        Text(getWarningMessage())
                            .font(.caption2)
                            .foregroundColor(.accessibleSecondary)
                    }
                    
                    Spacer()
                }
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(AccessibleColors.warningBackground)
                .cornerRadius(8)
            }
        }
        .padding(16)
        .background(.regularMaterial)
        .cornerRadius(12)
        .onAppear {
            withAnimation(.easeInOut(duration: 1.0)) {
                animatedOldProgress = impact.oldProgress
            }
            
            withAnimation(.easeInOut(duration: 1.0).delay(0.3)) {
                animatedNewProgress = impact.newProgress
            }
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel("Impact preview showing changes to your goal")
        .accessibilityValue(getAccessibilityDescription())
    }
    
    private func getWarningMessage() -> String {
        if impact.dailyTargetChange > 50 {
            return "Daily target increased significantly. Consider adjusting your deadline."
        } else if impact.progressChange < -0.2 {
            return "Progress percentage will decrease substantially."
        } else if impact.daysRemainingChange < -30 {
            return "Deadline moved much closer. Ensure the new timeline is realistic."
        } else {
            return "Review the changes carefully before saving."
        }
    }
    
    private func getAccessibilityDescription() -> String {
        let progressChange = Int((impact.newProgress - impact.oldProgress) * 100)
        let dailyChange = impact.newDailyTarget - impact.oldDailyTarget
        
        var description = "Progress will change by \(progressChange) percentage points. "
        description += String(format: "Daily target will change by %.2f %@. ", dailyChange, currency)
        
        if impact.isPositiveChange {
            description += "This is a positive change."
        } else {
            description += "This change requires attention."
        }
        
        return description
    }
}

// MARK: - Supporting Views
struct ChangeRow: View {
    let title: String
    let oldValue: String
    let newValue: String
    let isPositive: Bool
    
    var body: some View {
        HStack {
            Text(title)
                .font(.caption)
                .foregroundColor(.accessibleSecondary)
            
            Spacer()
            
            HStack(spacing: 8) {
                Text(oldValue)
                    .font(.caption)
                    .strikethrough()
                    .foregroundColor(.accessibleSecondary)
                
                Image(systemName: "arrow.right")
                    .font(.caption2)
                    .foregroundColor(.accessibleSecondary)
                
                Text(newValue)
                    .font(.caption)
                    .fontWeight(.medium)
                    .foregroundColor(isPositive ? AccessibleColors.success : AccessibleColors.warning)
            }
        }
    }
}

struct CircularProgressView: View {
    let progress: Double
    let size: CGFloat
    let lineWidth: CGFloat
    let color: Color
    
    var body: some View {
        ZStack {
            Circle()
                .stroke(color.opacity(0.2), lineWidth: lineWidth)
            
            Circle()
                .trim(from: 0, to: progress)
                .stroke(
                    color,
                    style: StrokeStyle(lineWidth: lineWidth, lineCap: .round)
                )
                .rotationEffect(.degrees(-90))
        }
        .frame(width: size, height: size)
    }
}

#Preview {
    VStack(spacing: 20) {
        // Positive change example
        ImpactPreviewCard(
            impact: GoalImpact(
                oldProgress: 0.45,
                newProgress: 0.60,
                oldDailyTarget: 75.0,
                newDailyTarget: 50.0,
                oldDaysRemaining: 120,
                newDaysRemaining: 150,
                oldTargetAmount: 5000,
                newTargetAmount: 5000,
                significantChange: true
            ),
            currency: "USD"
        )
        
        // Negative change example
        ImpactPreviewCard(
            impact: GoalImpact(
                oldProgress: 0.60,
                newProgress: 0.40,
                oldDailyTarget: 50.0,
                newDailyTarget: 100.0,
                oldDaysRemaining: 150,
                newDaysRemaining: 90,
                oldTargetAmount: 5000,
                newTargetAmount: 8000,
                significantChange: true
            ),
            currency: "USD"
        )
    }
    .padding()
}

File: android/app/src/main/java/com/xax/CryptoSavingsTracker/presentation/components/ImpactPreviewCard.kt
package com.xax.CryptoSavingsTracker.presentation.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.automirrored.filled.TrendingDown
import androidx.compose.material.icons.automirrored.filled.TrendingUp
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.xax.CryptoSavingsTracker.domain.model.GoalImpact
import com.xax.CryptoSavingsTracker.presentation.theme.AccessibleGreen
import com.xax.CryptoSavingsTracker.presentation.theme.InfoBlue
import com.xax.CryptoSavingsTracker.presentation.theme.WarningOrange
import java.text.NumberFormat
import java.util.Locale
import kotlin.math.abs

/**
 * Card showing the impact preview of goal changes.
 * Matches iOS ImpactPreviewCard for feature parity.
 */
@Composable
fun ImpactPreviewCard(
    impact: GoalImpact,
    currency: String = "USD",
    modifier: Modifier = Modifier
) {
    val accessibilityDescription = remember(impact) {
        buildAccessibilityDescription(impact, currency)
    }

    Card(
        modifier = modifier
            .fillMaxWidth()
            .semantics { contentDescription = accessibilityDescription },
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.5f)
        )
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            // Header
            ImpactHeader(isPositive = impact.isPositiveChange)

            // Progress Comparison
            ProgressComparisonRow(impact = impact)

            // Key Changes
            KeyChangesSection(impact = impact, currency = currency)

            // Warning if significant negative change
            if (!impact.isPositiveChange && impact.significantChange) {
                SignificantChangeWarning(impact = impact)
            }
        }
    }
}

@Composable
private fun ImpactHeader(isPositive: Boolean) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Icon(
            imageVector = if (isPositive) Icons.AutoMirrored.Filled.TrendingUp else Icons.AutoMirrored.Filled.TrendingDown,
            contentDescription = null,
            tint = if (isPositive) AccessibleGreen else WarningOrange,
            modifier = Modifier.size(24.dp)
        )

        Column {
            Text(
                text = "Impact Preview",
                style = MaterialTheme.typography.titleMedium
            )
            Text(
                text = if (isPositive) "Positive change" else "Requires attention",
                style = MaterialTheme.typography.labelSmall,
                color = if (isPositive) AccessibleGreen else WarningOrange
            )
        }

        Spacer(modifier = Modifier.weight(1f))
    }
}

@Composable
private fun ProgressComparisonRow(impact: GoalImpact) {
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text(
            text = "Progress",
            style = MaterialTheme.typography.labelLarge
        )

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceEvenly,
            verticalAlignment = Alignment.CenterVertically
        ) {
            // Before
            ProgressIndicator(
                label = "Before",
                progress = impact.oldProgress,
                color = InfoBlue
            )

            // Arrow
            Icon(
                imageVector = Icons.AutoMirrored.Filled.ArrowForward,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(24.dp)
            )

            // After
            ProgressIndicator(
                label = "After",
                progress = impact.newProgress,
                color = if (impact.isPositiveChange) AccessibleGreen else WarningOrange
            )
        }
    }
}

@Composable
private fun ProgressIndicator(
    label: String,
    progress: Double,
    color: Color
) {
    var animatedProgress by remember { mutableFloatStateOf(0f) }

    LaunchedEffect(progress) {
        animatedProgress = progress.toFloat()
    }

    val animatedValue by animateFloatAsState(
        targetValue = animatedProgress,
        animationSpec = tween(durationMillis = 1000),
        label = "progress"
    )

    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        CircularProgressIndicator(
            progress = animatedValue,
            size = 60.dp,
            strokeWidth = 6.dp,
            color = color
        )

        Text(
            text = "${(progress * 100).toInt()}%",
            style = MaterialTheme.typography.labelMedium
        )
    }
}

@Composable
private fun CircularProgressIndicator(
    progress: Float,
    size: Dp,
    strokeWidth: Dp,
    color: Color
) {
    Box(
        modifier = Modifier.size(size),
        contentAlignment = Alignment.Center
    ) {
        Canvas(modifier = Modifier.size(size)) {
            // Background circle
            drawArc(
                color = color.copy(alpha = 0.2f),
                startAngle = 0f,
                sweepAngle = 360f,
                useCenter = false,
                style = Stroke(width = strokeWidth.toPx(), cap = StrokeCap.Round)
            )

            // Progress arc
            drawArc(
                color = color,
                startAngle = -90f,
                sweepAngle = 360f * progress,
                useCenter = false,
                style = Stroke(width = strokeWidth.toPx(), cap = StrokeCap.Round)
            )
        }
    }
}

@Composable
private fun KeyChangesSection(impact: GoalImpact, currency: String) {
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        if (abs(impact.targetAmountChange) > 0.01) {
            ChangeRow(
                title = "Target Amount",
                oldValue = formatCurrency(impact.oldTargetAmount, currency),
                newValue = formatCurrency(impact.newTargetAmount, currency),
                isPositive = impact.targetAmountChange > 0
            )
        }

        if (abs(impact.dailyTargetChange) > 0.01) {
            ChangeRow(
                title = "Daily Target",
                oldValue = formatCurrency(impact.oldDailyTarget, currency),
                newValue = formatCurrency(impact.newDailyTarget, currency),
                isPositive = impact.dailyTargetChange < 0 // Less daily target is better
            )
        }

        if (impact.daysRemainingChange != 0) {
            ChangeRow(
                title = "Days Remaining",
                oldValue = "${impact.oldDaysRemaining} days",
                newValue = "${impact.newDaysRemaining} days",
                isPositive = impact.daysRemainingChange > 0 // More days is better
            )
        }
    }
}

@Composable
private fun ChangeRow(
    title: String,
    oldValue: String,
    newValue: String,
    isPositive: Boolean
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = title,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = oldValue,
                style = MaterialTheme.typography.labelSmall.copy(
                    textDecoration = TextDecoration.LineThrough
                ),
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )

            Icon(
                imageVector = Icons.AutoMirrored.Filled.ArrowForward,
                contentDescription = null,
                modifier = Modifier.size(12.dp),
                tint = MaterialTheme.colorScheme.onSurfaceVariant
            )

            Text(
                text = newValue,
                style = MaterialTheme.typography.labelMedium,
                color = if (isPositive) AccessibleGreen else WarningOrange
            )
        }
    }
}

@Composable
private fun SignificantChangeWarning(impact: GoalImpact) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(WarningOrange.copy(alpha = 0.1f))
            .padding(horizontal = 12.dp, vertical = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Icon(
            imageVector = Icons.Default.Warning,
            contentDescription = null,
            tint = WarningOrange,
            modifier = Modifier.size(16.dp)
        )

        Column {
            Text(
                text = "Significant Change",
                style = MaterialTheme.typography.labelMedium
            )
            Text(
                text = getWarningMessage(impact),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }

        Spacer(modifier = Modifier.weight(1f))
    }
}

private fun getWarningMessage(impact: GoalImpact): String {
    return when {
        impact.dailyTargetChange > 50 -> "Daily target increased significantly. Consider adjusting your deadline."
        impact.progressChange < -0.2 -> "Progress percentage will decrease substantially."
        impact.daysRemainingChange < -30 -> "Deadline moved much closer. Ensure the new timeline is realistic."
        else -> "Review the changes carefully before saving."
    }
}

private fun formatCurrency(amount: Double, currency: String): String {
    val format = NumberFormat.getCurrencyInstance(Locale.US)
    format.currency = java.util.Currency.getInstance(currency)
    return format.format(amount)
}

private fun buildAccessibilityDescription(impact: GoalImpact, currency: String): String {
    val progressChange = ((impact.newProgress - impact.oldProgress) * 100).toInt()
    val dailyChange = impact.newDailyTarget - impact.oldDailyTarget

    return buildString {
        append("Impact preview showing changes to your goal. ")
        append("Progress will change by $progressChange percentage points. ")
        append("Daily target will change by ${formatCurrency(dailyChange, currency)}. ")
        if (impact.isPositiveChange) {
            append("This is a positive change.")
        } else {
            append("This change requires attention.")
        }
    }
}
